client.go

```diff
@@ -0,0 +1,400 @@
+)
+
+// getLogs in a thread.
+func (s *service) getLogs(ctx context.Context, id thread.ID, pid peer.ID) ([]thread.LogInfo, error) {
+	req := &pb.GetLogsRequest{
+		Header: &pb.GetLogsRequest_Header{
+			From: &pb.ProtoPeerID{ID: s.threads.host.ID()},
+		},
+		ThreadID: &pb.ProtoThreadID{ID: id},
+	}
+
+	log.Debugf("getting thread %s logs from %s...", id.String(), pid.String())
+
+	cctx, cancel := context.WithTimeout(ctx, reqTimeout)
+	defer cancel()
+	conn, err := s.dial(cctx, pid, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+	client := pb.NewThreadsClient(conn)
+	reply, err := client.GetLogs(cctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugf("received %d logs from %s", len(reply.Logs), pid.String())
+
+	lgs := make([]thread.LogInfo, len(reply.Logs))
+	for i, l := range reply.Logs {
+		lgs[i] = logFromProto(l)
+	}
+
+	return lgs, nil
+}
+
+// pushLog to a peer.
+func (s *service) pushLog(ctx context.Context, id thread.ID, lid peer.ID, pid peer.ID) error {
+	lg, err := s.threads.store.LogInfo(id, lid)
+	if err != nil {
+		return err
+	}
+
+	lreq := &pb.PushLogRequest{
+		Header: &pb.PushLogRequest_Header{
+			From: &pb.ProtoPeerID{ID: s.threads.host.ID()},
+		},
+		ThreadID: &pb.ProtoThreadID{ID: id},
+		Log:      logToProto(lg),
+	}
+
+	log.Debugf("pushing log %s to %s...", lg.ID.String(), pid.String())
+
+	cctx, cancel := context.WithTimeout(ctx, reqTimeout)
+	defer cancel()
+	conn, err := s.dial(cctx, pid, grpc.WithInsecure())
+	if err != nil {
+		return fmt.Errorf("dial %s failed: %s", pid.String(), err)
+	}
+	client := pb.NewThreadsClient(conn)
+	_, err = client.PushLog(cctx, lreq)
+	return err
+}
+
+// records maintains an ordered list of records from multiple sources.
+type records struct {
+	sync.RWMutex
+	m map[cid.Cid]thread.Record
+	s []thread.Record
+}
+
+// newRecords creates an instance of records.
+func newRecords() *records {
+	return &records{
+		m: make(map[cid.Cid]thread.Record),
+		s: make([]thread.Record, 0),
+	}
+}
+
+// List all records.
+func (r *records) List() []thread.Record {
+	r.RLock()
+	defer r.RUnlock()
+	return r.s
+}
+
+// Store a record.
+func (r *records) Store(key cid.Cid, value thread.Record) {
+	r.Lock()
+	defer r.Unlock()
+	if _, ok := r.m[key]; ok {
+		return
+	}
+	r.m[key] = value
+	r.s = append(r.s, value)
+}
+
+// getRecords from log addresses.
+func (s *service) getRecords(
+	ctx context.Context,
+	id thread.ID,
+	lid peer.ID,
+	offsets map[peer.ID]cid.Cid,
+	limit int,
+) (map[peer.ID][]thread.Record, error) {
+	lgs := make(map[peer.ID]thread.LogInfo)
+	var err error
+	for lid := range offsets {
+		lgs[lid], err = s.threads.store.LogInfo(id, lid)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	pblgs := make([]*pb.GetRecordsRequest_LogEntry, 0)
+	for _, lg := range lgs {
+		pblgs = append(pblgs, &pb.GetRecordsRequest_LogEntry{
+			LogID:  &pb.ProtoPeerID{ID: lg.ID},
+			Offset: &pb.ProtoCid{Cid: offsets[lg.ID]},
+			Limit:  int32(limit),
+		})
+	}
```

Comment:
Maybe this loop work can be done in the above loop? Since `offsets` is a map, it shouldn't have some repeated-values problem which might affect `append`. (+ `make` with `len(offsets)` cap)

Reply:
yep, doing now